### PR TITLE
Table view performance

### PR DESCRIPTION
**Pull request by antares1470**
### Description of the Change

A minor performance improvement in the table view by cutting down the excess number of threads being created when switching between graphs

### Alternate Designs

Am open to suggestions for other ways to check for a thread interruption after each line of code. What I have works but it looks a little messy so would prefer something cleaner if it exists

### Why Should This Be In Core?

Table view is in core and this directly affects this

### Benefits

Table view now snappier when switching between graphs

### Applicable Issues

#409

### CODE REVIEW

**Comment by arcturus2** [comment]
I know I said to you that its more important to get it to work than look pretty...baaaht after looking at this again I don't like this change sorry @antares1470.

I think though the change and what you are trying to achieve with reducing the table threads is important so I will see if I can suggest an alternative; though I don't have anything right now.

**Comment by formalhaut69** [comment]
@antares1470 the tableview looks visually snappier now. I got a null pointer exception when swapping between the graphs. 

java.lang.NullPointerException
	at javafx.controls/javafx.scene.control.skin.CellSkinBase.access$100(CellSkinBase.java:56)
	at javafx.controls/javafx.scene.control.skin.CellSkinBase$StyleableProperties$1.isSettable(CellSkinBase.java:166)
	at javafx.controls/javafx.scene.control.skin.CellSkinBase$StyleableProperties$1.isSettable(CellSkinBase.java:161)
	at javafx.graphics/javafx.scene.CssStyleHelper.transitionToState(CssStyleHelper.java:666)
	at javafx.graphics/javafx.scene.Node.doProcessCSS(Node.java:9654)
	at javafx.graphics/javafx.scene.Node.access$900(Node.java:398)
	at javafx.graphics/javafx.scene.Node$1.doProcessCSS(Node.java:471)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSSImpl(NodeHelper.java:192)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSSImpl(ParentHelper.java:93)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSS(ParentHelper.java:63)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Parent.java:1366)
	at javafx.graphics/javafx.scene.Parent.access$400(Parent.java:79)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Parent.java:125)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(ParentHelper.java:98)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSSImpl(ControlHelper.java:63)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSS(ControlHelper.java:55)
	at javafx.controls/javafx.scene.control.Control.doProcessCSS(Control.java:886)
	at javafx.controls/javafx.scene.control.Control.access$000(Control.java:83)
	at javafx.controls/javafx.scene.control.Control$1.doProcessCSS(Control.java:89)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.processCSSImpl(ControlHelper.java:67)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Parent.java:1397)
	at javafx.graphics/javafx.scene.Parent.access$400(Parent.java:79)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Parent.java:125)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(ParentHelper.java:98)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSSImpl(ControlHelper.java:63)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSS(ControlHelper.java:55)
	at javafx.controls/javafx.scene.control.Control.doProcessCSS(Control.java:886)
	at javafx.controls/javafx.scene.control.Control.access$000(Control.java:83)
	at javafx.controls/javafx.scene.control.Control$1.doProcessCSS(Control.java:89)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.processCSSImpl(ControlHelper.java:67)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Parent.java:1397)
	at javafx.graphics/javafx.scene.Parent.access$400(Parent.java:79)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Parent.java:125)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(ParentHelper.java:98)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Parent.java:1397)
	at javafx.graphics/javafx.scene.Parent.access$400(Parent.java:79)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Parent.java:125)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(ParentHelper.java:98)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Parent.java:1397)
	at javafx.graphics/javafx.scene.Parent.access$400(Parent.java:79)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Parent.java:125)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(ParentHelper.java:98)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSSImpl(ControlHelper.java:63)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.superProcessCSS(ControlHelper.java:55)
	at javafx.controls/javafx.scene.control.Control.doProcessCSS(Control.java:886)
	at javafx.controls/javafx.scene.control.Control.access$000(Control.java:83)
	at javafx.controls/javafx.scene.control.Control$1.doProcessCSS(Control.java:89)
	at javafx.controls/com.sun.javafx.scene.control.ControlHelper.processCSSImpl(ControlHelper.java:67)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.graphics/javafx.scene.Node.processCSS(Node.java:9536)
	at javafx.graphics/javafx.scene.Node.processCSS(Node.java:9529)
	at javafx.graphics/javafx.scene.Node.processCSS(Node.java:9529)
	at javafx.graphics/javafx.scene.Node.processCSS(Node.java:9529)
	at javafx.graphics/javafx.scene.Node.processCSS(Node.java:9529)
	at javafx.graphics/javafx.scene.Scene.doCSSPass(Scene.java:569)
	at javafx.graphics/javafx.scene.Scene.access$3400(Scene.java:172)
	at javafx.graphics/javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2477)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Toolkit.java:412)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:411)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:438)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:519)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:499)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(QuantumToolkit.java:492)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(QuantumToolkit.java:320)
[catch] at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	at java.base/java.lang.Thread.run(Thread.java:834)

**Comment by aldebaran30701** [comment]
I get an exception on this branch, wheras on master I don't.

```
java.lang.ArrayIndexOutOfBoundsException: Index 4100 out of bounds for length 4096
	at au.gov.asd.tac.constellation.visual.opengl.utilities.glyphs.GlyphRectangleBuffer.addRectImage(GlyphRectangleBuffer.java:224)
	at au.gov.asd.tac.constellation.visual.opengl.utilities.glyphs.GlyphManagerBI.lambda$renderTextAsLigatures$5(GlyphManagerBI.java:421)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.visual.opengl.utilities.glyphs.GlyphManagerBI.renderTextAsLigatures(GlyphManagerBI.java:415)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.NodeLabelBatcher.bufferTopLabel(NodeLabelBatcher.java:243)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.NodeLabelBatcher.fillTopLabels(NodeLabelBatcher.java:204)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.NodeLabelBatcher.createBatch(NodeLabelBatcher.java:157)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GraphRenderable.lambda$getChangeProcessor$3(GraphRenderable.java:162)
	at au.gov.asd.tac.constellation.utilities.visual.VisualProcessor.lambda$processChangeSet$1(VisualProcessor.java:296)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.utilities.visual.VisualProcessor.lambda$processChangeSet$2(VisualProcessor.java:295)
	at java.base/java.util.HashMap.forEach(HashMap.java:1336)
	at au.gov.asd.tac.constellation.utilities.visual.VisualProcessor.processChangeSet(VisualProcessor.java:293)
	at au.gov.asd.tac.constellation.utilities.visual.VisualProcessor.lambda$update$0(VisualProcessor.java:206)
[catch] at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by antares1470** [comment]
Converting to draft while I sort out all of these issues.
@formalhaut69 @aldebaran30701 Do you recall what you did to generate the exceptions? I had done a fair bit of testing with switching between graphs but I never encountered either of the exception you found

**Comment by aldebaran30701** [comment]
@antares1470 I created two graphs. one i did sphere builder with 100 elements, the other with 1000 elements. I had the table view opened and switched between each graph a few times. I then closed the table view and it caused the exception. I also had it on another occasion where it happened as soon as i switched the graph once, without having to close the table view.

**Comment by formalhaut69** [comment]
> 
> 
> Converting to draft while I sort out all of these issues.
> @formalhaut69 @aldebaran30701 Do you recall what you did to generate the exceptions? I had done a fair bit of testing with switching between graphs but I never encountered either of the exception you found

@antares1470 I just kept swapping between the graphs for maybe 20-30 seconds. The memory usage just kept going up so I wanted to see what would happen.

**Comment by antares1470** [comment]
Thanks @aldebaran30701 and @formalhaut69

**Comment by antares1470** [comment]
@formalhaut69 made a suggestion of using ScheduledFutures which was used as a solution to a previous bug (#210). So far it looks good in testing and certainly looks cleaner than the previous solution. Thoughts @arcturus2 

As a side note, I haven't yet been able to replicate the exceptions mentioned above (both with this solution and the previous), so I'm wondering what the specific circumstances might be for these occurring (e.g. in the null pointer case I could get Constellation lagging, but never actually triggering the null pointer).

**Comment by formalhaut69** [comment]
> 
> 
> @formalhaut69 made a suggestion of using ScheduledFutures which was used as a solution to a previous bug (#210). So far it looks good in testing and certainly looks cleaner than the previous solution. Thoughts @arcturus2?
> 
> As a side note, I haven't yet been able to replicate the exceptions mentioned above (both with this solution and the previous), so I'm wondering what the specific circumstances might be for these occurring (e.g. in the null pointer case I could get Constellation lagging, but never actually triggering the null pointer).

@antares1470 I had a quick test with your latest commit and I can still reproduce the nullpointer unfortunately. Glad that you were able to repurpose that code though.

**Comment by aldebaran30701** [comment]
I'd like to say that the null pointer @formalhaut69 found was also reproducible on the current master. 
With the current state of this PR, I could replicate the array out of bounds issue I pasted above.

**Comment by aldebaran30701** [comment]
@antares1470 I'd just like to mention that I was working on #321 and had made it multi-threaded and during testing I got the array out of bounds exception. I'm not sure if it is a multi-threaded thing or if it is a bug in master. I'll be soon to test master to verify this. Just hope I didn't send you on a wild goose chase if that was the case!

File: CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java [hunk 1]
```diff
@@ -193,6 +200,8 @@ public TableViewPane(final TableViewTopComponent parent) {
         };
         this.selectedProperty = table.getSelectionModel().selectedItemProperty();
         selectedProperty.addListener(tableSelectionListener);
+        
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(1);
```
Comment (arcturus2) [inline CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:204]: ❤️

**Comment by antares1470** [comment]
Thanks @aldebaran30701 for those findings. When I looked at the stack traces, it didn't exactly make sense as to why the exceptions would only be occurring in my branch (given nothing I changed popped up in the stack trace), so I'm glad to hear that it issue isn't specific to my branch.

Consequently, I think this PR is ready for review

**Pull request merged**